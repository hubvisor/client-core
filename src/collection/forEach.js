import reduce from './reduce'

/**
 * A forEach function that can be applied to both arrays or objects.
 * Follows `Array.forEach` for semantics except for thisArg ! @see {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach}
 * @param {array|object} source The object or array to be iterated on.
 * @param {function} eachFn A callback to be called for each (value, key) pair of the input value with (value, key, source) arguments.
 * @returns {undefined}
 */
export default function forEach (source, eachFn) {
  reduce(source, (last, value, key, source) => eachFn(value, key, source), undefined)
}
