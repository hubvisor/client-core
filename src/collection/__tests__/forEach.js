import forEach from '../forEach'

describe('reduce', () => {
  describe('Array', () => {
    test('empty', () => {
      const cb1 = jest.fn()
      const cb2 = jest.fn()
      const empty = []

      forEach(empty, cb1)
      empty.forEach(cb2)

      expect(cb1.mock.calls).toEqual(cb2.mock.calls)
    })

    test('non-empty', () => {
      const cb1 = jest.fn()
      const cb2 = jest.fn()
      const array = [ 1, 2, 3 ]

      forEach(array, cb1)
      array.forEach(cb2)

      expect(cb1.mock.calls).toEqual(cb2.mock.calls)
    })
  })

  describe('Object', () => {
    const object = { foo: 42, bar: 10 }

    test('empty', () => {
      const empty = {}
      const cb = jest.fn()

      forEach(empty, cb)

      expect(cb.mock.calls).toEqual([])
    })

    test('non-empty', () => {
      const empty = {}
      const cb = jest.fn()

      forEach(empty, cb)

      expect(cb).toMatchSnapshot()
    })
  })
})
