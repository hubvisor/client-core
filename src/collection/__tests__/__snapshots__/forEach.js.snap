// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduce Object non-empty 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      42,
      "foo",
      Object {
        "bar": 10,
        "foo": 42,
      },
    ],
    Array [
      10,
      "bar",
      Object {
        "bar": 10,
        "foo": 42,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
